lib/filewatch/read_mode/handlers/read_file.rb

```diff
@@ -19,8 +19,10 @@ def handle_specifically(watched_file)
           watched_file.listener.eof
           watched_file.file_close
           key = watched_file.sincedb_key
-          sincedb_collection.reading_completed(key)
-          sincedb_collection.clear_watched_file(key)
+          if sincedb_collection.get(key)
```

Comment:
is it possible that by line 22 the key still exists, but then in 23 or 24 it doesn't anymore? maybe we can just change the reading_complete and clear_watched_file to tolerate the absence of the key, as we don't have a synchornized way to access this collection

Reply:
No it's not possible. The problem is generated from the same thread, with this sequence of actions:
- https://github.com/logstash-plugins/logstash-input-file/blob/15d3eba2dbd723243134affea3eb09888a7b3240/lib/filewatch/read_mode/handlers/read_file.rb#L12 
- https://github.com/logstash-plugins/logstash-input-file/blob/15d3eba2dbd723243134affea3eb09888a7b3240/lib/filewatch/sincedb_collection.rb#L30
- https://github.com/logstash-plugins/logstash-input-file/blob/15d3eba2dbd723243134affea3eb09888a7b3240/lib/filewatch/sincedb_collection.rb#L185
- this triggers the invocation of serializer, which the discover our file was read > 1ms ago, so put it in the list of `expired_keys` https://github.com/logstash-plugins/logstash-input-file/blob/15d3eba2dbd723243134affea3eb09888a7b3240/lib/filewatch/sincedb_collection.rb#L216
  - https://github.com/logstash-plugins/logstash-input-file/blob/15d3eba2dbd723243134affea3eb09888a7b3240/lib/filewatch/sincedb_record_serializer.rb#L25
- this delete the key from sincedb_collection https://github.com/logstash-plugins/logstash-input-file/blob/15d3eba2dbd723243134affea3eb09888a7b3240/lib/filewatch/sincedb_collection.rb#L219

Reply:
great stuff Andrea, makes sense to me.